src/libexpr/eval.cc

```diff
@@ -347,6 +347,16 @@ void EvalState::allowPath(const StorePath & storePath)
         rootFS2->allowPrefix(CanonPath(store->toRealPath(storePath)));
 }
 
+void EvalState::allowClosure(const StorePath & storePath)
+{
+    if (!rootFS.dynamic_pointer_cast<AllowListSourceAccessor>()) return;
+
+    StorePathSet closure;
+    store->computeFSClosure(storePath, closure);
```

Comment:
Computing the closure from scratch could cause a performance problem.
`n` calls to `allowClosure` would be O(n * closure size), which may be significantly larger than O(total allowed closure size).
The latter can be achieved by computing the closure of paths that aren't in an allowed closure yet, i.e. checking the references recursively and skipping ones that have been of the closure of any preceding `allowClosure` call.

Reply:
The cost of the call to `computeFSClosure()` is likely to be trivial compared to the actual build of the derivation we're importing from.